src/components/AvailabilityGrid.tsx

```diff
@@ -0,0 +1,155 @@
+'use client'
+import React, { useState, useEffect } from 'react'
+import { generateTimeRange } from '@/utils/timeUtils'
+
+interface GridProps {
+  earliestTime: string
+  latestTime: string
+  isAvailable: boolean // Determines if the grid is selectable
+  daysOfWeek: string[]
+}
+
+const Grid = ({
+  earliestTime,
+  latestTime,
+  isAvailable,
+  daysOfWeek,
+}: GridProps) => {
+  // Generate time array for row headings
+  const timeArray = generateTimeRange(earliestTime, latestTime)
+
+  // Grid dimensions
+  const dimensions = {
+    width: daysOfWeek.length || 1, // default to 1 if daysOfWeek is empty
+    height: timeArray.length,
+  }
+
+  // Function to Initialize and populate an empty 2d array
+  const initialGrid = () => {
+    return Array(dimensions.height)
+      .fill(null)
+      .map(() => Array(dimensions.width).fill(false))
+  }
+
+  const [grid, setGrid] = useState(initialGrid)
+  const [isSelecting, setIsSelecting] = useState(false)
+
+  // Update grid dimensions when daysOfWeek changes
+  useEffect(() => {
+    setGrid(initialGrid())
+  }, [daysOfWeek.length, timeArray.length])
+
+  // Function is called when the mouse is pressed down on a cell
+  const handleMouseDown = (rowIndex: number, colIndex: number) => {
+    if (!isAvailable) return
+    console.log(`Mouse Down: Row ${rowIndex}, Col ${colIndex}`)
+    setIsSelecting(true)
+    toggleCell(rowIndex, colIndex)
+  }
+
+  /* 
+    Function is called when the mouse enters a cell while 
+    the mouse button is pressed 
+  */
+  const handleMouseEnter = (rowIndex: number, colIndex: number) => {
+    if (!isAvailable) return
+    if (isSelecting) {
+      console.log(`Mouse Enter: Row ${rowIndex}, Col ${colIndex}`)
+      toggleCell(rowIndex, colIndex)
+    }
+  }
+
+  // Function is called when the mouse button is released
+  const handleMouseUp = () => {
+    console.log('Mouse Up')
+    setIsSelecting(false)
+  }
+
+  // Function toggles the value of a cell
+  const toggleCell = (rowIndex: number, colIndex: number) => {
+    console.log(`Toggling Cell: Row ${rowIndex}, Col ${colIndex}`)
+    const newGrid = [...grid]
```

Comment:
Not a fan of this because it is a shallow copy of the grid to set a value.

This isn't necessarily more readable but this might be better:
  const newGrid = grid.map((row, rIndex) =>
    rIndex === rowIndex
      ? row.map((cell, cIndex) => (cIndex === colIndex ? !cell : cell))
      : row
  );

  setGrid(newGrid);

Reply:
Keep in mind for if there is a performance issue.